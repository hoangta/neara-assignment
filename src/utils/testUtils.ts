import { MFA } from "../types/mfa";

export const mockStore = () => {
  jest.mock("../stores/mfaStore", () =>
    jest.fn().mockImplementation(() => ({
      mfas: mockMFAs,
      createMFA: jest.fn(),
      switchMFAs: jest.fn(),
    }))
  );
};

export const mockSystemTime = (date: Date) => {
  jest.useFakeTimers("modern");
  jest.setSystemTime(date);
};

export const mockSVGElementAnimate = () => {
  const circleAnimateSpy = jest.fn();
  SVGElement.prototype.animate = circleAnimateSpy;
  return circleAnimateSpy;
};

export const mockMFAs: MFA[] = [
  {
    provider: "Test Provider",
    code: "012345",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 2),
  },
  {
    provider: "Test Provider 2",
    code: "543210",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 3),
  },
];
