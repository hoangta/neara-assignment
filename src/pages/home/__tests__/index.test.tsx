import React from "react";
import { fireEvent, render } from '@testing-library/react'
import Home from '..'
import MFAStore from "../../../stores/mfaStore";
import { mockSVGElementAnimate, mockMFAs } from '../../../utils/testUtils';

mockSVGElementAnimate();

describe('Pages/Home', () => {
    const mockStore = new MFAStore();
    const onCreateSpy = jest.fn();

    it('should render correctly', () => {
        const { asFragment } = render(<Home store={mockStore} onCreate={onCreateSpy} />);
        expect(asFragment()).toMatchSnapshot();
    })

    it('should handle events correctly', () => {
        const { queryByTestId, getByTestId } = render(<Home store={mockStore} onCreate={onCreateSpy} />);
        //Not editing
        expect(queryByTestId('moveup_button_0')).not.toBeInTheDocument();
        expect(queryByTestId('movedown_button_0')).not.toBeInTheDocument();
        //Editing
        fireEvent.click(getByTestId('edit_button'))
        expect(getByTestId('moveup_button_0')).toBeInTheDocument();
        expect(getByTestId('movedown_button_0')).toBeInTheDocument();
        //Create
        fireEvent.click(getByTestId('add_button'));
        expect(onCreateSpy).toBeCalledTimes(1);
        //Switch MFA
        fireEvent.click(getByTestId('moveup_button_0'));
        expect(mockStore.switchMFAs).toBeCalledWith(0, -1);
        fireEvent.click(getByTestId('movedown_button_0'));
        expect(mockStore.switchMFAs).toBeCalledWith(0, 1);
    })
})