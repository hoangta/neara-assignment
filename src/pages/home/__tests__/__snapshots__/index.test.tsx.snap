// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/Home should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="home_page"
  >
    <div
      class="container"
    >
      <button
        class="button"
        data-testid="edit_button"
      >
        Edit
      </button>
      <div
        class="title"
      >
        Tokens
      </div>
      <button
        class="button"
        data-testid="add_button"
        style="font-size: 25px;"
      >
        ï¼‹
      </button>
    </div>
    <div>
      <div
        class="container"
      >
        <img
          alt="Test Provider"
          class="image"
          src="www.testurl.com"
        />
        <div
          class="detail"
        >
          <div
            class="provider"
          >
            Test Provider
          </div>
          <div
            class="code"
          >
            012 345
          </div>
        </div>
        <div
          class="indicatorContainer"
        >
          <svg
            class="indicatorItem"
          >
            <circle
              class="circle"
              cx="20"
              cy="20"
              fill="none"
              r="14"
              stroke="gray"
              stroke-dasharray="87.96459430051421"
              stroke-dashoffset="87.96459430051421"
              stroke-width="2"
            />
          </svg>
          <div
            class="countdownLabel"
          >
            86400
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <img
          alt="Test Provider 2"
          class="image"
          src="www.testurl.com"
        />
        <div
          class="detail"
        >
          <div
            class="provider"
          >
            Test Provider 2
          </div>
          <div
            class="code"
          >
            543 210
          </div>
        </div>
        <div
          class="indicatorContainer"
        >
          <svg
            class="indicatorItem"
          >
            <circle
              class="circle"
              cx="20"
              cy="20"
              fill="none"
              r="14"
              stroke="gray"
              stroke-dasharray="87.96459430051421"
              stroke-dashoffset="87.96459430051421"
              stroke-width="2"
            />
          </svg>
          <div
            class="countdownLabel"
          >
            172800
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <img
          alt="Test Provider 3"
          class="image"
          src="www.testurl.com"
        />
        <div
          class="detail"
        >
          <div
            class="provider"
          >
            Test Provider 3
          </div>
          <div
            class="code"
          >
            567 890
          </div>
        </div>
        <div
          class="indicatorContainer"
        >
          <svg
            class="indicatorItem"
          >
            <circle
              class="circle"
              cx="20"
              cy="20"
              fill="none"
              r="14"
              stroke="gray"
              stroke-dasharray="87.96459430051421"
              stroke-dashoffset="87.96459430051421"
              stroke-width="2"
            />
          </svg>
          <div
            class="countdownLabel"
          >
            172800
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <img
          alt="Test Provider 4"
          class="image"
          src="www.testurl.com"
        />
        <div
          class="detail"
        >
          <div
            class="provider"
          >
            Test Provider 4
          </div>
          <div
            class="code"
          >
            098 765
          </div>
        </div>
        <div
          class="indicatorContainer"
        >
          <svg
            class="indicatorItem"
          >
            <circle
              class="circle"
              cx="20"
              cy="20"
              fill="none"
              r="14"
              stroke="gray"
              stroke-dasharray="87.96459430051421"
              stroke-dashoffset="87.96459430051421"
              stroke-width="2"
            />
          </svg>
          <div
            class="countdownLabel"
          >
            172800
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
